File: notifications-api-python/src/storage.py
from datetime import datetime

from models import Notification, SENT, FAILED, RETRY_PENDING
from segmenter import min_sms_segments

notifications = []
next_id = 1


def add_notification(target_channels, message):
    global next_id
    n = Notification(next_id, target_channels, message)
    if any(c.get("type") == "sms" for c in target_channels):
        n.smsSegments = min_sms_segments(message)
    next_id += 1
    notifications.append(n)
    return n


def get_all():
    return notifications


def find_by_id(nid):
    for n in notifications:
        if n.id == nid:
            return n
    return None


def seed():
    global next_id
    notifications.clear()
    next_id = 1

    n1 = add_notification([{"type": "email", "value": "alice@example.com"}], "Welcome to the platform")
    n1.status = SENT
    n1.attempts = 1
    n1.lastAttemptAt = datetime.now().isoformat()
    n1.lastError = "[email] accepted for delivery"

    add_notification([{"type": "sms", "value": "12345"}], "Short number")

    n3 = add_notification([{"type": "push", "value": "device-abc"}], "Your ride is here")
    n3.status = FAILED
    n3.attempts = 1
    n3.lastAttemptAt = datetime.now().isoformat()
    n3.lastError = "[push] device token rejected"

    add_notification(
        [
            {"type": "email", "value": "bob@example.com"},
            {"type": "sms", "value": "+15551234567"},
        ],
        "2FA code 4242",
    )

    n5 = add_notification(
        [
            {"type": "sms", "value": "+15559876543"},
            {"type": "push", "value": "device-xyz"},
            {"type": "email", "value": "carol@example.com"},
        ],
        "Order shipped",
    )
    n5.status = RETRY_PENDING
    n5.attempts = 2
    n5.lastAttemptAt = datetime.now().isoformat()
    n5.lastError = "[sms] temporary outage, retry later"

File: notifications-api-typescript/src/storage.ts
import {
  Notification,
  TargetChannel,
  SENT,
  FAILED,
  RETRY_PENDING,
} from "./models.js";
import { minSmsSegments } from "./segmenter.js";

const notifications: Notification[] = [];
let nextId = 1;

export function addNotification(
  targetChannels: TargetChannel[],
  message: string
): Notification {
  const n = new Notification(nextId++, targetChannels, message);
  if (targetChannels.some((c) => c.type === "sms")) {
    n.smsSegments = minSmsSegments(message);
  }
  notifications.push(n);
  return n;
}

export function getAll(): Notification[] {
  return notifications;
}

export function findById(id: number): Notification | undefined {
  return notifications.find((n) => n.id === id);
}

export function seed(): void {
  notifications.length = 0;
  nextId = 1;

  const n1 = addNotification(
    [{ type: "email", value: "alice@example.com" }],
    "Welcome to the platform"
  );
  n1.status = SENT;
  n1.attempts = 1;
  n1.lastAttemptAt = new Date();
  n1.lastError = "[email] accepted for delivery";

  addNotification([{ type: "sms", value: "12345" }], "Short number");

  const n3 = addNotification(
    [{ type: "push", value: "device-abc" }],
    "Your ride is here"
  );
  n3.status = FAILED;
  n3.attempts = 1;
  n3.lastAttemptAt = new Date();
  n3.lastError = "[push] device token rejected";

  addNotification(
    [
      { type: "email", value: "bob@example.com" },
      { type: "sms", value: "+15551234567" },
    ],
    "2FA code 4242"
  );

  const n5 = addNotification(
    [
      { type: "sms", value: "+15559876543" },
      { type: "push", value: "device-xyz" },
      { type: "email", value: "carol@example.com" },
    ],
    "Order shipped"
  );
  n5.status = RETRY_PENDING;
  n5.attempts = 2;
  n5.lastAttemptAt = new Date();
  n5.lastError = "[sms] temporary outage, retry later";
}
